src/main/java/com/google/firebase/auth/ListUsersPage.java

```diff
@@ -0,0 +1,260 @@
+    return currentBatch.getUsers();
+  }
+
+  private static class UserIterable implements Iterable<ExportedUserRecord> {
+
+    private final ListUsersPage startingPage;
+
+    UserIterable(@NonNull ListUsersPage startingPage) {
+      this.startingPage = checkNotNull(startingPage, "starting page must not be null");
+    }
+
+    @Override
+    @NonNull
+    public Iterator<ExportedUserRecord> iterator() {
+      return new UserIterator(startingPage);
+    }
+
+    /**
+     * An {@code Iterator} that cycles through user accounts, one at a time. It buffers the
+     * last retrieved batch of user accounts in memory. The {@code maxResults} parameter is an
+     * upper bound on the batch size.
+     */
+    private static class UserIterator implements Iterator<ExportedUserRecord> {
+
+      private ListUsersPage currentPage;
+      private List<ExportedUserRecord> batch;
+      private int index = 0;
+
+      private UserIterator(ListUsersPage startingPage) {
+        setCurrentPage(startingPage);
+      }
+
+      @Override
+      public boolean hasNext() {
+        if (index == batch.size()) {
+          if (currentPage.hasNextPage()) {
+            setCurrentPage(currentPage.getNextPage());
+          } else {
+            return false;
+          }
+        }
+
+        return index < batch.size();
+      }
+
+      @Override
+      public ExportedUserRecord next() {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+        return batch.get(index++);
+      }
+
+      @Override
+      public void remove() {
+        throw new UnsupportedOperationException("remove operation not supported");
+      }
+
+      private void setCurrentPage(ListUsersPage page) {
+        this.currentPage = checkNotNull(page);
+        this.batch = ImmutableList.copyOf(page.getValues());
+        this.index = 0;
+      }
+    }
+  }
+
+  /**
+   * Represents a source of user data that can be queried to load a batch of users.
+   */
+  interface UserSource {
+    @NonNull
+    ListUsersResult fetch(int maxResults, String pageToken);
+  }
+
+  static class DefaultUserSource implements UserSource {
+
+    private final FirebaseUserManager userManager;
+
+    DefaultUserSource(FirebaseUserManager userManager) {
+      this.userManager = checkNotNull(userManager, "user manager must not be null");
+    }
+
+    @Override
+    public ListUsersResult fetch(int maxResults, String pageToken) {
+      try {
+        DownloadAccountResponse response = userManager.listUsers(maxResults, pageToken);
+        ImmutableList.Builder<ExportedUserRecord> builder = ImmutableList.builder();
+        if (response.getUsers() != null) {
+          for (DownloadAccountResponse.User user : response.getUsers()) {
+            builder.add(new ExportedUserRecord(user));
+          }
+        }
+        String nextPageToken = response.getPageToken() != null
+            ? response.getPageToken() : END_OF_LIST;
```

Comment:
Have you though about using "null" to mark the final page token and just validating that the end user never passes it in?

Reply:
Users should be able to pass `null` here, which stands for the first page.